restapi/src/main/java/io/stargate/web/docsapi/service/query/rules/NegateFilters.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright The Stargate Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.stargate.web.docsapi.service.query.rules;
+
+import com.bpodgursky.jbool_expressions.Expression;
+import com.bpodgursky.jbool_expressions.Not;
+import com.bpodgursky.jbool_expressions.options.ExprOptions;
+import com.bpodgursky.jbool_expressions.rules.Rule;
+import com.bpodgursky.jbool_expressions.rules.RuleList;
+import com.bpodgursky.jbool_expressions.rules.RulesHelper;
+import io.stargate.web.docsapi.service.query.FilterExpression;
+import java.util.Collections;
+import java.util.List;
+
+public class NegateFilters extends Rule<Expression<FilterExpression>, FilterExpression> {
+
+  private static final RuleList<FilterExpression> rules =
+      new RuleList<>(Collections.singletonList(new NegateFilters()));
+
+  @Override
+  public Expression<FilterExpression> applyInternal(
+      Expression<FilterExpression> input, ExprOptions<FilterExpression> options) {
+    Not<FilterExpression> negated = (Not<FilterExpression>) input;
+    FilterExpression child = (FilterExpression) negated.getChildren().get(0);
+
+    return child.negate();
+  }
+
+  @Override
+  protected boolean isApply(Expression<FilterExpression> input) {
+    if (!(input instanceof Not)) {
+      return false;
+    }
+
+    Not<FilterExpression> negated = (Not<FilterExpression>) input;
```

Comment:
That said I would really like to have an INT test that tackles this situation!

Reply:
The intent behind this rule is to convert something like `Not(GT(1))` into `LTE(1)`. So, the rule only accepts `Not` expression that have a `FilterExpression` as a direct child.

It should work both for CNF and DNF as those forms have `NOT`s pushed down to the predicate (variable) level.

This rule is a bit useless during expression simplification. Granted it can simplify some cases (e.g. `And(Not(GT(1))`), but it will not be able to eliminate `NOT`s completely, e.g. in `Not(And(GT(1), LT(2))`. Conversion to CNF or DNF is required to push the negation to predicate level in that case (simplification alone is not sufficient).